### PR TITLE
Value usage improvements

### PR DESCRIPTION
**Pull request by sdroege**
Fixes https://github.com/gtk-rs/glib/issues/223 and https://github.com/gtk-rs/glib/issues/228

### CODE REVIEW

**Comment by EPashkin** [comment]
and maybe even more generalized by adding function parameter

**Comment by sdroege** [comment]
What would the signature of the function you think of look like?

**Comment by EPashkin** [comment]
I thought about something like this, but this version always allocate vector in stack,
as I don't see way to fill uninitialized vector.
```
fn func<'a, T,U, V: Fn(&U)->T>(f:V, values: &[U], stack: &'a mut[T], heap: &'a mut Vec<T>) -> &'a[T] {
  if values.len() <= stack.len() {
      for (i, arg) in values.iter().enumerate() {
          stack[i] = f(arg);
      }
      &stack[0..values.len()]
  } else {
      heap.reserve(stack.len());
      for (i, arg) in values.iter().enumerate() {
          heap[i] = f(arg);
      }
      heap.as_slice()
  }
}
```

**Comment by sdroege** [comment]
Note that this wouldn't cover both variants. The one in object.rs puts the instance argument as first, additional element into the vector.

File: src/object.rs [hunk 1]
```diff
@@ -736,24 +737,35 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
...
-            c_args.push(ptr::read(instance.to_glib_none().0));
-            for arg in args {
-                c_args.push(ptr::read(arg.to_glib_none().0));
-            }
+            let mut v_args: Vec<Value>;
+            let mut s_args: [Value; 10] = mem::zeroed();
+            let args = if args.len() < 10 {
+                s_args[0] = Value::from(self);
+                for (i, arg) in args.iter().enumerate() {
+                    s_args[i] = arg.to_value();
```
Comment (EPashkin) [inline src/object.rs]: This not replace s_args[0] ?

**Comment by EPashkin** [comment]
Yes, you right. `values` can be changed to `ExactSizeIterator`.

**Comment by EPashkin** [comment]
.. and second version use `std::iter::once` and `chain`.

**Comment by EPashkin** [comment]
Maybe generalizing not worth memory or speed-wise, so I don't insist.
Also you can try https://crates.io/crates/smallvec as it seems doing almost same.

**Comment by sdroege** [comment]
It seemed like too much to depend on an external crate just for 4 lines of code :) but for refactoring the right solution would be a data type and not a function

Thanks for your other comments, I'll clean it up tomorrow

File: src/object.rs [hunk 2]
```diff
@@ -736,24 +738,33 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
...
+            let mut v_args: Vec<Value>;
+            let mut s_args: [Value; 10] = mem::zeroed();
+            let args = if args.len() < 10 {
+                for (i, arg) in iter::once(&(self as &ToValue)).chain(args).enumerate() {
+                    s_args[i] = arg.to_value();
+                }
+                &s_args[0..args.len()+1]
+            } else {
+                v_args = Vec::with_capacity(args.len() + 1);
+                for arg in iter::once(&(self as &ToValue)).chain(args) {
```
Comment (sdroege) [inline src/object.rs]: I'm not sure how to get rid of the `&(self as &ToValue)` disaster here. Any suggestions?

Comment (EPashkin) [reply src/object.rs]: Why it disaster?
By way, I proposed use `iter::once` for function, if used directly then old variant looks normal (after index fix and `with_capacity` on second branch)

Comment (sdroege) [reply src/object.rs]: It looks ugly but works :) do you prefer it like this or the previous version?

Comment (EPashkin) [reply src/object.rs]: IMHO almost all same as long as it works

File: src/value.rs [hunk 3]
```diff
@@ -93,6 +93,7 @@ use gobject_ffi;
 /// (e.g. numeric types) don't.
 ///
 /// See the [module documentation](index.html) for more details.
+#[repr(C)]
 pub struct Value(gobject_ffi::GValue);
```
Comment (sdroege) [inline src/value.rs]: Let's get this in too then, this is not a new requirement for this PR but already before Closure and other places casted between glib::Value and gobject_ffi::GValue for preventing copies.

File: src/object.rs (hunk 2 above)
Comment (sdroege) [reply src/object.rs]: 👍 Let's keep it as now then

File: src/value.rs (hunk 3 above)
Comment (EPashkin) [reply src/value.rs]: IMHO better be `repr(internal)` but now I can't find it at all.

**Comment by EPashkin** [comment]
@sdroege Thanks
👍 for merge.

**Comment by GuillaumeGomez** [comment]
Let's merge then. Thanks to both of you!

**Pull request merged**